lib/chrontext/src/web.rs

```diff
@@ -0,0 +1,61 @@
+use std::sync::Arc;
+
+use axum::response::Html;
+use axum::Form;
+use axum::{self, Router};
+use axum::{extract::State, routing::get};
+
+use oxigraph::sparql::{results::QueryResultsFormat, QueryResults, QuerySolutionIter};
+use oxrdf::Variable;
+use serde::Deserialize;
+
+use crate::sparql_database::SparqlQueryable;
+
+#[derive(Clone)]
+struct AppState {
+    sparql_engine: Arc<(dyn SparqlQueryable)>,
```

Comment:
Is there any reason I might be missing as to maybe implementing this over `engine` instead of `SparqlQueryables`?

I think this is only letting you run queries on some underlying sparql database, and I'm worried about how this might interact with the virtualization stuff.

the `Engine::query` return type `(DataFrame, HashMap<String, RDFNodeType>, Vec<Context>)` is pretty complicated, so I need to take a deeper dive into how this works

Reply:
Should be implemented over engine, yes. 

The `DataFrame, HashMap<String, RDFNodeType>` representation is a column-based encoding of a result.
For each variable, there is a column. The map holds the RDF type of the column. In case the variable has multiple types, there is a Struct-column with multiple columns for that type. 

There is https://github.com/DataTreehouse/maplib/blob/main/lib/representation/src/polars_to_rdf.rs which maps the df and types to a row based result of the kind we need here. Might need a bit of cleaning up though, but should be fairly well tested.